frontend/projects/upgrade/src/app/features/dashboard/home/components/experiment-query-result/experiment-query-result.component.ts

```diff
@@ -261,13 +250,11 @@ export class ExperimentQueryResultComponent implements OnInit, OnDestroy {
   getLevelName(levelId: string): string {
     // TODO: look for factors at index apart from 0
     let levelName;
-    this.experiment.partitions.forEach((decisionPoint) => {
-      decisionPoint.factors.forEach((factor) => {
-        factor.levels.forEach((level) => {
-          if (level.id === levelId) {
-            levelName = level.name;
-          }
-        });
+    this.experiment.factors.forEach((factor) => {
+      factor.levels.forEach((level) => {
```

Comment:
> this is a minor thing but I think a similar searching pattern like this comes up a lot and should be a hint that we're doing too much work: One thing is that `forEach` isn't the right tool for the job, as it will continue iterating after it finds a match, so`find` might be best.
> 
> But it would be better to take a step back altogether, because this would do all of these iterations every time `getLevelName()` is invoked also. Ultimately it's not a huge deal as there won't be 1000 factors and levels to sift through, but if you cache a flat reference map of levels at startup (or in ngrx store as soon as the experiment is upserted...) then it's just a straight lookup each time, which would be ideal:
> 
> for example (this is untested code for illustration purposes, use at your own risk!)
> 
> ```ts
> /**
>  * What we want is a flat map of "id: Level":
>  * { 
>  *   'abc1': Level,
>  *   'd6f3': Level,
>  *   'a2b5': Level
>  * }
>  */
> type LevelsMap = { [key: string] : ExperimentLevel };
> 
> this.levelsMap: LevelsMap = {};
> 
> ngOnInit() {
>   // ...do things
>   this.levels = this.createLevelsMap(this.experiment.factors); // make a flat lookup map one time
>   // ...do more things
> }
> 
> createLevelsMap(factors: ExperimentFactor[]): LevelsMap {
>   return factors.reduce((levelsMap, factor) => {
>       factor.levels.forEach((level) => {
>          levelsMap[level.id] = level
>       })
>       
>       return levelsMap;
>   }, {})
> }
> 
> // then this accessor is simple:
> getLevelName(levelId: string): string {
>   const level = this.levelsMap[levelId];
>   
>   return level?.name || "";
> }
> ```

Okay, I will look into this. Thanks :)